numpy/ma/core.py

```diff
@@ -205,6 +205,31 @@ class MaskError(MAError):
     min_filler.update([(np.float128, +np.inf)])
 
 
+def _recursive_fill_value(dtype, f):
+    """
+    Recursively produce a fill value for `dtype`, calling f on scalar dtypes
+    """
+    if dtype.names:
+        vals = tuple(_recursive_fill_value(dtype[name], f) for name in dtype.names)
+        return np.array(vals, dtype=dtype)[()]  # decay to void scalar from 0d
+    elif dtype.subdtype:
+        subtype, shape = dtype.subdtype
+        subval = _recursive_fill_value(subtype, f)
+        return np.full(shape, subval)
+    else:
+        return f(dtype)
+
+
+def _get_dtype_of(obj):
+    """ Convert the argument for *_fill_value into a dtype """
```

Comment:
This is a pretty weird overload, but we're stuck with it. In particular, `default_fill_value('i')` and `default_fill_value(np.dtype('i'))` do not mean the same thing.

This PR changes the behaviour of `...([1, 2, 3])`, which previously used `np.object_`, but now builds the array and returns `np.int32`.